### PR TITLE
Save on shutdown

### PR DESCRIPTION
**Pull request by sandy081**
Implement #194491

- Save on shutdown

### CODE REVIEW

File: src/vs/platform/quickinput/browser/commandsQuickAccess.ts [hunk 1]
```diff
@@ -334,17 +344,15 @@ export class CommandsHistory extends Disposable {
 			return;
 		}
 
-		this.configuredCommandsHistoryLength = CommandsHistory.getConfiguredCommandHistoryLength(this.configurationService);
+		this.configuredCommandsHistoryLength = this.getConfiguredLength();
 
-		if (CommandsHistory.cache && CommandsHistory.cache.limit !== this.configuredCommandsHistoryLength) {
-			CommandsHistory.cache.limit = this.configuredCommandsHistoryLength;
-
-			CommandsHistory.saveState(this.storageService);
```
Comment (TylerLeonhardt) [inline src/vs/platform/quickinput/browser/commandsQuickAccess.ts:342]: with the removal of this line, I don't see when `save` (fka `saveState`) is called in the Standalone case. Not entirely sure what that did in that case, but if it did _something_ this PR causes state saving to not happen.

File: src/vs/platform/quickinput/browser/commandsQuickAccess.ts [hunk 2]
```diff
@@ -301,17 +308,20 @@ interface ICommandsQuickAccessConfiguration {
...
+
+	readonly _serviceBrand: undefined;
 
 	static readonly DEFAULT_COMMANDS_HISTORY_LENGTH = 50;
 
 	private static readonly PREF_KEY_CACHE = 'commandPalette.mru.cache';
 	private static readonly PREF_KEY_COUNTER = 'commandPalette.mru.counter';
-
-	private static cache: LRUCache<string, number> | undefined;
 	private static counter = 1;
```
Comment (TylerLeonhardt) [inline src/vs/platform/quickinput/browser/commandsQuickAccess.ts:319]: nit: this can be a "not static" member. Not a blocking comment.

**Review by bpasero — CHANGES_REQUESTED** [review]
Can we not just save when the `onWillSaveState` indicates a shutdown? See for example how suggest memory only saves state on shutdown to reduce spam:

https://github.com/microsoft/vscode/blob/29db3adb03d9f805bbc7c324242ff42bc4d1b089/src/vs/editor/contrib/suggest/browser/suggestMemory.ts#L242-L246

File: src/vs/platform/quickinput/browser/commandsQuickAccess.ts (hunk 1 above)
Comment (sandy081) [reply src/vs/platform/quickinput/browser/commandsQuickAccess.ts:342]: I belieive in standalone case we do not have persistent storage. So saving state is a noop in this case and reload will always starts fresh. But changing to @bpasero suggestion will handle this

**Comment by sandy081** [comment]
@bpasero Thats a good suggestion. Changed.

File: src/vs/platform/quickinput/browser/commandsQuickAccess.ts [hunk 3]
```diff
@@ -327,6 +328,11 @@ export class CommandsHistory extends Disposable {
 
 	private registerListeners(): void {
 		this._register(this.configurationService.onDidChangeConfiguration(e => this.updateConfiguration(e)));
+		this._register(this.storageService.onWillSaveState(e => {
+			if (e.reason === WillSaveStateReason.SHUTDOWN) {
+				this.saveState();
```
Comment (bpasero) [inline src/vs/platform/quickinput/browser/commandsQuickAccess.ts:333]:
```suggestion
				// Commands history is very dynamic and so we limit impact
				// on storage to only save on shutdown. This helps reduce
				// the overhead of syncing this data across machines.
				this.saveState();
```

File: src/vs/platform/quickinput/browser/commandsQuickAccess.ts [hunk 4]
```diff
@@ -338,8 +344,6 @@ export class CommandsHistory extends Disposable {
 
 		if (CommandsHistory.cache && CommandsHistory.cache.limit !== this.configuredCommandsHistoryLength) {
 			CommandsHistory.cache.limit = this.configuredCommandsHistoryLength;
```
Comment (bpasero) [inline src/vs/platform/quickinput/browser/commandsQuickAccess.ts:346]: Either we have to set `isDirty` here or trigger a save to ensure we element count is reduced in storage.

File: src/vs/platform/quickinput/browser/commandsQuickAccess.ts [hunk 5]
```diff
@@ -310,6 +310,7 @@ export class CommandsHistory extends Disposable {
 
 	private static cache: LRUCache<string, number> | undefined;
 	private static counter = 1;
+	private static isDirty = false;
```
Comment (bpasero) [inline src/vs/platform/quickinput/browser/commandsQuickAccess.ts:313]: I associate `dirty` more with editors, maybe `shouldSaveState`?

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/platform/quickinput/browser/commandsQuickAccess.ts (hunk 5 above)
Comment (sandy081) [reply src/vs/platform/quickinput/browser/commandsQuickAccess.ts:313]: Renamed to `hasChanges`

File: src/vs/platform/quickinput/browser/commandsQuickAccess.ts (hunk 4 above)
Comment (sandy081) [reply src/vs/platform/quickinput/browser/commandsQuickAccess.ts:346]: Done

**Review by bpasero — APPROVED** [review]

**Pull request merged**